🚀 1단계 buildMethod를 Widget으로 분리하기

jambottle opened the pull request:
## 요구사항

### 1차 기능 구현

<img width="426" alt="flutter-cart-sample-step1" src="https://user-images.githubusercontent.com/72926450/199851295-4b3e25c6-aebb-4b1b-b5e0-3322a7256447.png">

- [x] buildMethod로 구현된 코드들을 Stateless Widget으로 구현
- [x] 클래스의 네이밍은 자유롭게 하되, 의미 파악이 명확하도록 작성
  - `ShopInfo`: 가게 프로필 (가게명, 썸네일)
  - `CartItem`: 장바구니 상품 카드
  - `OrderContinue`: '더 담으러 가기' 버튼
  - `OrderComplete`: '배달 주문하기' 버튼 → 클릭 시, 결제 화면(Payment)으로 이동할 것으로 기대
  - `PaymentInfo`: 결제 정보 (주문금액, 배달팁)

### 2차 기능 구현

- [x] Intl 라이브러리를 활용해서 가격에 Formatter 적용
- [x] 데이터를 생성자로 전달받아서 다른 음식점, 다른 메뉴도 표시할 수 있도록 수정

### 기타

- [x] GestureDetector와 Button 차이 경험

jambottle committed:
rename: Create an assets/images folder and Move images to the folder

jambottle committed:
refactor: Add ShopProfile widget to replace _buildStoreName method

jambottle committed:
refactor: Add ProductCard widget to replace _buildMenu method

jambottle committed:
refactor: Add ProductMore widget to replace _buildAddMore method

jambottle committed:
refactor: Add PaymentInfo widget to replace _buildBilling method

jambottle committed:
refactor: Separate OrderButton widget from CartScreen widget

jambottle committed:
style: Use trailing commas for automatic code formatting

jambottle committed:
chore(deps): Install intl

jambottle committed:
feat: Add Formatter class and getWon method for number formatting

jambottle committed:
feat: Pass mock data through constructor (ShopProfile/ProductCard)

Kiboom commented on lib/utils/formatter.dart:
```diff
@@ -0,0 +1,12 @@
+import 'package:intl/intl.dart';
+
+class Formatter {
+  static String getWon(double number) {
```
`Formatter`로 포맷 코드를 공통화 하셨군요! 좋은 아이디어 입니다!👍

Kiboom commented on lib/utils/formatter.dart:
```diff
@@ -0,0 +1,12 @@
+
+class Formatter {
+  static String getWon(double number) {
+    var f = NumberFormat.currency(
```
* 변경될 일이 없는 값은 `final`로 선언해주시는 걸 권장드려요 😀 

* 가독성을 위해 이니셜보다는 단어로 변수명을 만드는 것을 권장드립니다-!

Kiboom commented on lib/widgets/product_card.dart:
```diff
@@ -0,0 +1,87 @@
+                child: ClipRRect(
+                  borderRadius: BorderRadius.circular(12),
+                  child: Image.asset(
+                    productData['thumbnail']!,
```
Map에 접근할 때는 `!`를 사용하는 것보다는 `??`를 사용함으로써 
키가 없는 상황을 대비하는 것을 권장드려요!

```dart
ex)
productData['thumbnail'] ?? ''
```

더 좋은 방법은, Map이 아닌 **class** 형태로 productData를 정의하는 것입니다.

필드를 선언하거나 접근하거나 필드명을 변경할 때, 
Map은 필드명을 일일이 타이핑 해줘야 하고 실수할 위험성도 있어요!

반면 class로 데이터를 정의하면,
IDE를 통해 더 빠르고 안전하게 필드에 접근하고 관리할 수 있습니다!

Kiboom commented on lib/widgets/product_card.dart:
```diff
@@ -0,0 +1,87 @@
+import '../utils/formatter.dart';
+
+class ProductCard extends StatelessWidget {
+  final Map<String, String> productData;
```
`productData`는 외부에 공개될 필요가 없는 필드이기 때문에 
private 필드로 만들어보면 어떨까요?

Kiboom commented on lib/widgets/shop_profile.dart:
```diff
@@ -0,0 +1,40 @@
+import 'package:flutter/material.dart';
+
+class ShopProfile extends StatelessWidget {
+  final Map<String, String> shopData;
```
`shopData`도 외부에 공개될 필요가 없는 필드이기 때문에 
private 필드로 만들어도 좋을 것 같아요!

Kiboom commented on lib/widgets/product_more.dart:
```diff
@@ -0,0 +1,35 @@
+import 'package:flutter/material.dart';
+
+class ProductMore extends StatelessWidget {
```
위젯을 별도의 파일로 잘 분리해주셨네요!👍

다만 `ProductMore`는 `cart_screen.dart`에서만 사용될 것 같아요!

`ProductMore`를 `cart_screen.dart` 파일에서만 사용할 수 있게 하고,
다른 파일에서는 못 쓰게 하려면 어떻게 해야할까요?

다음 링크를 참고해서 수정해보시겠어요?🙂

https://github.com/next-step/flutter-cart-sample/pull/1#pullrequestreview-1086807145

jambottle replied:
확실히 `cart_screen.dart`에서만 사용될 것 같은 위젯들에 한해 `part - part of`를 적용해보았습니다! (https://github.com/next-step/flutter-cart-sample/pull/21/commits/d420d6aa9b42a35cab127f387fed1ae34ae0fd53)

jambottle replied:
𝑸. Dart 언어에서는 private임을 선언할 때 이름 앞에 밑줄(`_`)을 붙이는 규칙이 있다고 알고 있는데,
만약 `part - part of`를 적용한 private 위젯의 경우에도 이름 앞에 밑줄을 붙여주는 게 맞는 표현일까요? 🤔

jambottle replied:
𝑸. Dart 언어에서는 private임을 선언할 때 이름 앞에 밑줄(`_`)을 붙이는 규칙이 있다고 알고 있는데,
이번에 `part - part of`를 적용한 private 위젯들의 경우에도 이름 앞에 밑줄을 붙여주는 게 맞는 표현일까요? 🤔

Kiboom replied:
네 맞습니다! 클래스명 앞에 `_`를 붙여주면 
`해당 파일 안에서만 private하게 접근 가능한 클래스`를 만들 수 있습니다! 😀

Kiboom replied:
정리하면, 변수명에 `_`를 붙이는 경우는 주로 다음과 같아요.

```
• 클래스 내의 멤버 변수나 메소드를 private으로 만들고 싶을 때
• 클래스를 파일 내에서 private하게 사용하고 싶을 때 
```

주의할 점은, 로컬 변수에는 `_`를 붙이지 않는 것이 좋습니다 !
읽는 이로 하여금 `이 변수는 멤버 변수이구나!`라고 착각하게 할 수 있거든요 🙊

[참고 문서](https://dart.dev/guides/language/effective-dart/style#dont-use-a-leading-underscore-for-identifiers-that-arent-private)

Kiboom commented on lib/cart_screen.dart:
```diff
@@ -10,6 +16,18 @@ class CartScreen extends StatefulWidget {
+      'thumbnail': 'assets/images/chicken.png',
+      'name': '케이준 크리스피 치킨',
+      'description': '찜 & 리뷰 약속 시, 서비스 음료 제공 (콜라)',
+    };
```
데이터를 정의할 때 Map보다는 class 형태로 정의해보면 어떨까요?

필드를 선언하거나 접근하거나 필드명을 변경할 때, 
Map은 필드명을 일일이 타이핑 해줘야 하고 실수할 위험성도 있어요!

반면 class로 데이터를 정의하면,
IDE를 통해 더 빠르고 안전하게 필드에 접근하고 관리할 수 있습니다!

Kiboom reviewed (changes requested):
첫 과제인데 잘 진행해주셨어요! 🎉👏

단순히 위젯을 분리하는 것 뿐만 아니라,
공통적으로 쓰이는 로직들은 한 곳에 잘 모아주셨네요! (ex. `Formatter`)

위젯 분리할 때 공통화도 함께 신경 쓰지 않으면 
중복 코드가 많이 발생할 수 있는데, 이 부분 잘 처리해주셨습니다! 💯

몇 가지 팁과 피드백 남겨드렸는데,
이 부분도 추가로 수정해보시면 좋을 것 같아요 ☺️

jambottle committed:
refactor: Make 3 child widgets private with part - part of directives

jambottle committed:
rename: Rename 4 child widgets to better serve their purpose

jambottle committed:
feat: Add 2 data models (Shop/Product)

jambottle committed:
feat: Use 2 data models for mock data instead of Map type
- Replace null assertion operator(with if-null operator(??)
- Use Offstage widget for conditional rendering of Text widget

jambottle committed:
feat: Update 2 data models (Shop/Product)
- Change parameter type of getWon method from double to int

jambottle committed:
feat: Add 1 data model (Payment)

jambottle committed:
feat: Pass mock data through constructor (PaymentInfo)

jambottle commented on lib/cart_screen.dart:
```diff
@@ -10,6 +21,25 @@ class CartScreen extends StatefulWidget {
   @override
   Widget build(BuildContext context) {
+    /* Mock Data */
+    final shopData = Shop(
+      '가깝고도 먼치킨 석촌호수점',
+      thumbnail: 'assets/images/chickenCartoonImage.jpg',
+      deliveryPrice: 3000,
+    );
+
+    final productData = Product(
+      '케이준 크리스피 치킨',
+      description: '찜 & 리뷰 약속 시, 서비스 음료 제공 (콜라)',
+      thumbnail: 'assets/images/chicken.png',
+      originalPrice: 18000,
+    );
+
+    final paymentData = Payment(
+      originalPrice: productData.originalPrice,
+      deliveryPrice: shopData.deliveryPrice,
+    );
```
Map으로 정의했던 데이터들을 **class** 형태로 다시 정의해보았습니다!
말씀해주신 대로 더 빠르고 안전하게 필드에 접근하고 관리할 수 있었어요.

개인적으로 JavaScript의 **object** 형태로 데이터를 정의하는 것에 좀 더 익숙한 편인데,
Dart/Flutter에서는 class로 데이터를 정의하는 것에 좀 더 익숙해져 봐야겠습니다..!! 🤔

jambottle committed:
feat: Pass mock data through constructor (OrderComplete)

jambottle commented on lib/widgets/cart_item.dart:
```diff
@@ -0,0 +1,88 @@
+                child: ClipRRect(
+                  borderRadius: BorderRadius.circular(12),
+                  child: Image.asset(
+                    _productData.thumbnail ?? '',
```
`!` 연산자 대신 `??` 연산자를 사용하여 값이 없는 상황을 대비해보았습니다!
값이 있을 거라고 확정 짓는 것보다 값이 없는 상황을 예방하는 것이 훨씬 견고한 것 같아요.

jambottle commented on lib/widgets/shop_info.dart:
```diff
@@ -0,0 +1,42 @@
+          ClipRRect(
+            borderRadius: BorderRadius.circular(12),
+            child: Image.asset(
+              _shopData.thumbnail ?? '',
```
`!` 연산자 대신 `??` 연산자를 사용하여 값이 없는 상황을 대비해보았습니다!
값이 있을 거라고 확정 짓는 것보다 값이 없는 상황을 예방하는 것이 훨씬 견고한 것 같아요.

Kiboom reviewed (approved):
리뷰 의도에 맞게 잘 반영해주셨네요-! 💯
[`part of`로 분리한 클래스들](https://github.com/next-step/flutter-cart-sample/commit/d420d6aa9b42a35cab127f387fed1ae34ae0fd53)만 private으로 바꿔보시면 좋을 듯 합니다 👍
고생 많으셨어요!!
